File: 100-Same Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque
class Solution(object):
    def isSameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        # BFS
        
        queue = deque([(p, q)])
        while queue:
            for _ in range(len(queue)):
                currP, currQ = queue.popleft()
                if currP and currQ:
                    if currP.val == currQ.val:
                        queue.append((currP.left, currQ.left))
                        queue.append((currP.right, currQ.right))
                    else:
                        return False
                elif currP or currQ:
                    return False
        return True
    
        # DFS
        
        def dfs(p, q):
            if not p and not q:
                return True
            if not p or not q:
                return False
            if p.val != q.val:
                return False
            return dfs(p.left, q.left) and dfs(p.right, q.right)
        
        return dfs(p, q)

File: JavaScript/100-Same Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {boolean}
 */
var isSameTree = function(p, q) {
    let dfs = (p, q) =>{
        if (!p && !q) return true;
        if (!p || !q) return false;
        if (p.val !== q.val) return false;
        return dfs(p.left, q.left) && dfs(p.right, q.right);
    }
    return dfs(p,q);
};